Torch Loader works with compact data.

alxmrs opened the pull request:
This PR depends on #211. Fixes #204.

Here is the WandB profile results: https://wandb.ai/m2lines/ocean-emulators/reports/Compaction-and-Vectorization-on-Data-Load-Time--VmlldzoxMjg1MDE4NQ

Conclusions: compaction is at least as fast as before, but often predictably faster.

alxmrs committed:
Introducing `DataSource`.

alxmrs committed:
Added compact data support to TorchLoader

alxmrs commented on src/ocean_emulators/datasets.py:
```diff
@@ -321,15 +322,27 @@ def _get_prognostic(self, x_index):
-        )
+        if src_in.is_compact:
+            data_in_np: np.ndarray = (
+                conditional_rearrange(
```
Once #165 is landed, we'll no longer need this `conditional_rearrange()` function.

alxmrs replied:
I found out that this is not quite true!

alxmrs committed:
Fixed pre-commit issues.

alxmrs committed:
Fixed broken tests: adjusted `validate_data` method and updated `conditional_rearrange` call for vectorized data.

alxmrs committed:
More correct data.

alxmrs commented on src/ocean_emulators/utils/data.py:
```diff
@@ -59,9 +94,12 @@ def map(
         ],
         *,
-        suffix: str,
+        suffix: str | None = None,
```
This is unrelated to compaction; I just thought it was nice-to-have.

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -37,20 +38,54 @@ class DataSource:
+        """Check if the data source is compact."""
+        return all(
+            v in d
+            for v in ["so", "uo", "vo", "thetao"]
```
Is there something in constants.py we could use instead of hard-coding these here?

alxmrs replied:
Found a better way.

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -37,20 +38,54 @@ class DataSource:
         *,
         prefix: str,
     ) -> Self:
         """Filter the data source to only include the specified variables."""
```
Maybe we could use a comment or expanded doc string about the behavior of this function (or class?) with compact data? Looking at the code it seems like it expects that callers might pass in var_names with some (but not all) levels? Presumably we require that the set of levels is the same across all vars?

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -59,9 +94,12 @@ def map(
+        if suffix is None:
+            suffix = func.__qualname__
+
         data, means, stds = func(self.data.copy(), self.means.copy(), self.stds.copy())
```
So given there's no change here, this means that map bodies (and map_data bodies) must be aware that data may or may not be compact, right? Maybe this is too big of a change, but would it be possible to somehow hide the compact-vs-not distinction as part of this class's API?

alxmrs replied:
I kind of have an idea of what that could look like, but I actually like the status quo right now. Some advice that I really like is to "pull ifs up and to push fors down"[1]. The way `is_compact` is used now means we change the `map` calls instead of having a more complicated `map` call that has two sections. Or, am I misunderstanding how this would change? In any case: the `validate_data` function is a good example of this, at a higher level, we do A vs B based on whether the data is compact. 


[1]: https://matklad.github.io/2023/11/15/push-ifs-up-and-fors-down.html

jder replied:
I see the value of the general guideline but at the moment the fact that this class mostly exposes (rather than hides) the compactness of the data means the conditional_rearrange call is duplicated 4 times outside this class and it's pretty hard for me to tell if it's covering all the places it needs to. Anyone that accesses the .data member might need be edited depending on compactness. (For example, do we need to change InferenceDataset.get_coords_dict?) As opposed to, say, the `filter` method which works the same regardless of the compactness of the data. 

Anyway, if you want to leave it that's ok for this PR, but I do feel like we will end up wanting to centralize the knowledge of what structure the data has down the road.

alxmrs replied:
You bright up really good points. I see how this API has it both ways (between `filter` and `map`). I'll see if I can spot a pattern in how `conditional_rearrange` is used and create a better abstraction. If I can't, I'm happy to reconsider this later on as we learn more. I am worried that it may be too early to generalized right now.

alxmrs replied:
Update! I had to change calls to `is_compact` just now to fix issues with test. I learned that `conditional_rearrange` isn't necessarily tied to compact data, and actually is tied to having a `lev` dimension. Now, the use of `is_compact` doesn't seem to have the external/internal difference. 

In any case, I think this shows that there is more discovery to be had in how these concepts are used.

jder replied:
What's the difference between `is_compact` and having a `lev` dimension?

alxmrs replied:
Fair point, I think these always co-occur in practice. When I replaced with `lev`, I think there was some corner case I uncovered, but it could have come from an unrelated change/cause.

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -150,6 +211,30 @@ def from_config(cls, cfg: TrainConfig | EvalConfig, *, use_dask: bool) -> Self:
+    data: xr.Dataset, pattern: str, except_dim="lev", concat_dim="variable"
+) -> xr.DataArray:
+    assert except_dim in pattern, f"{except_dim} must be in the pattern."
+    data_ = data.copy()
```
I don't think you need this copy.

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -150,6 +211,30 @@ def from_config(cls, cfg: TrainConfig | EvalConfig, *, use_dask: bool) -> Self:
 
+def conditional_rearrange(
+    data: xr.Dataset, pattern: str, except_dim="lev", concat_dim="variable"
+) -> xr.DataArray:
```
I think a doc comment would help with understandability here.

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -370,8 +455,11 @@ def with_lat_lon_coords(data: xr.Dataset) -> xr.Dataset:
     """Validate the data such that we have the correct format for training."""
+    if src.is_compact:
+        src_ = src.map_data(with_lat_lon_coords)
+        return compute_anomalies(src_, ("hfds_anomalies",))
```
Can we pull this from TensorMap rather than hard-coding?

jder reviewed (commented):
OK, sorry, I didn't get a chance to look carefully at everything but I have to run so leaving a few comments here and will return to this tonight or tomorrow.

alxmrs committed:
Better way to check if data is compact via standard metadata constant.

alxmrs committed:
Better docstring.

alxmrs committed:
Literate programming style docstring.

alxmrs committed:
Eliminated a copy.

alxmrs committed:
Not hard-coding; avoiding using tensor-map.

alxmrs committed:
Tests passing; fixed issues related to merge.
- new `is_compact` was bad; only testing the first 4 vars
- rearrange should not include time.
- time dimension is in a different order.

alxmrs committed:
Run ruff.

jder commented on src/ocean_emulators/datasets.py:
```diff
@@ -58,6 +58,8 @@ def __init__(
 
         self.hist = hist
+        self.src = src
+        data = src.data
```
self.src appears to be unused, and data is assigned 2 lines below, so I don't think we want either of these lines of code?

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -154,6 +226,52 @@ def from_config(cls, cfg: TrainConfig | EvalConfig, *, use_dask: bool) -> Self:
+        data[vars_with_dim]
+        .to_array()
+        .einops.rearrange(pattern, dask="allowed")
+        .drop_vars(concat_dim, errors="ignore")
```
Why do we drop_vars?

alxmrs replied:
It's because of a weird xarray implementation detail where the concat dim is considered both a dimension and a variable. If I don't call this, then it causes errors when concating IIRC.

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -38,20 +40,64 @@ class DataSource:
+    @cached_property
+    def is_compact(self) -> bool:
+        """Check if the data source is compact."""
+        standard_vars = list(DEFAULT_METADATA.keys())[:3]
```
Reading the first 3 keys of a dictionary intended for another purpose seems pretty fragile to me, and for example seems like it will break if we're training on the `thetao_surface` variable set since that will exclude so/uo. How about checking for `_lev_` in the data var names? That or else passing is_compact into the constructor instead?

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -38,20 +40,64 @@ class DataSource:
+                    parsed_var_names.append(mangled_var_name)
+                    continue
+                tokens = mangled_var_name.split("_")
+                var_name, level = tokens[0], int(tokens[1])
```
🐑 Should we be asserting `var_name == "lev"` here?

alxmrs replied:
I don't think so; `lev` will likely be a dimension on this var, but we want to know what the variable name will be.

jder replied:
Ah, right, sorry I misread.

jder reviewed (changes requested):
Thanks, generally looks good, left a few comments. What do you think about including a test that the compact data produces the same data as the non-compact for the torch loader?

alxmrs committed:
Clean up unused code from merge and overlooking useage.

alxmrs committed:
Better definition of `is_compact`; fix for `normalize_with`.

alxmrs committed:
More accurate test than `is_compact`.

alxmrs committed:
Update validate_data given the new API.

alxmrs commented:
> What do you think about including a test that the compact data produces the same data as the non-compact for the torch loader?

I really like this idea, and I'm looking into a way of doing it that doesn't make our pytest configuration too complicated. Do you have any ideas for this? I'm still research at the moment. I think your new config changes make this kind of thing much easier to do, but the single `data_source` fixture makes this type of combination really hard. Maybe, the simplest think to do would be to make each data source into its own fixture, and then combine them into a "super" fixture. If you like that approach, then I think it's best to do that in a new PR.

jder commented:
> > What do you think about including a test that the compact data produces the same data as the non-compact for the torch loader?

> I really like this idea, and I'm looking into a way of doing it that doesn't make our pytest configuration too complicated.

I think probably I would avoid injecting fixtures for this and build two train_configs (and data_sources) by calling those fixtures as functions (or breaking their core logic out into a separate function) and then using make_loader on both. If you want to do some refactoring here in a separate PR to make that easy, that makes sense to me -- do you want to do that first? (To be clear, I'm fine either way.)

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -38,20 +39,66 @@ class DataSource:
+            "_" not in str(v)
+            for d in [self.data, self.means, self.stds]
+            for v in d.keys()
+            if "anom" not in str(v)
```
This feels like this is going to bite us if we are doing this heuristically rather than having it be a property read from TensorMap, some clearer property of the data variables themselves (eg we mark them with `_lev_` in their name), or the compactness is passed in to the data source. For example, #224 in introduces some new data variables with underscores in their names.

alxmrs replied:
Reading the tensormap sources now, it looks like they also partition on `"_"` to parse level variables. I agree heuristics are fragile, but I think this one may be good enough for now, until we can replace everything with compact data.

jder replied:
Ok, I think given #224 has landed there could now be a `sea_surface_fraction` in the data keys. How about looking for `_[0-9]+$` or something if you don't want to do a larger refactoring? (And probably the same thing needs to happen in filter below.)

alxmrs replied:
Thanks, these are great ideas.

alxmrs committed:
WIP, trying to get tensor's aligned.

alxmrs committed:
Fix: removed vectorization, and now tests seem to pass.

alxmrs committed:
uv ruff

alxmrs committed:
Conditional rearrange is not needed for the boundary.

alxmrs commented:
> Anyway, if you want to leave it that's ok for this PR, but I do feel like we will end up wanting to centralize the knowledge of what structure the data has down the road.

This is a fair point, but I want to lean putting this off till later mainly because I hope that soon, we will just use compact data and not have to support the non-standard Zarr store in the first place.

alxmrs committed:
minimize diff.

alxmrs committed:
Added failing test to compare compact and flattened data.

alxmrs committed:
Fix: `conditional_rearrange` is now stable (preserves rough sorted order of variables).

alxmrs commented on src/ocean_emulators/utils/data.py:
```diff
@@ -156,6 +231,89 @@ def from_config(cls, cfg: TrainConfig | EvalConfig, *, use_dask: bool) -> Self:
+            np.arange(n_center + n_front, n_front + n_center + n_back),  # rest of vars
+        )
+    )
+    order_da = xr.DataArray(order, dims=concat_dim)
```
@jder The test you asked for revealed a bug that this new block of code fixes: This rearrange also changes the normal order of variables. I don't think it would have affected any ML code, but now we can be sure the compact data is exactly equal to the data before. 

Thanks thinking of writing a test for this!

alxmrs commented on tests/test_datasets.py:
```diff
@@ -393,6 +401,59 @@ def test_new_loaders__are_equal_to_v1_data_loader(train_config, loader_version):
 
 
+@pytest.mark.parametrize("data_source", ["remote-om4"], indirect=True)
+def test_compact_loader__equals_flat_loader(
```
This was my KISS approach to have both remote and compact data in a test. It involves a bit of boilerplate, but I am happy that I don't have to boil the ocean to test these two data sources.

jder replied:
It looks like this depends on the implementation details of `conftest.data_source`. If you don't want to refactor that code in order to reuse the relevant pieces, can you please put a comment here and in there pointing to the other copy and explaining how they share internals?

alxmrs replied:
A comment is a good idea; I decided to raise an assertion for stronger enforcement.

jder replied:
I meant more that the cache paths + names used have to match `conftest.data_source` so I still think a comment in each pointing to the other is a good idea.

Looking at this again, I don't think we need either `flat_data` or `compact_data` (and so don't need `_compact` either) since we only ever use their name to build a config?

alxmrs replied:
Very interesting! I see what you mean now; I'm computing a compact value that is, probably always, cached. Ok, I see what you mean now.

alxmrs committed:
Walrus operator to make array ops a little more readable.

alxmrs committed:
Handle anomaly vars when filtering compact data.

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -156,6 +234,90 @@ def from_config(cls, cfg: TrainConfig | EvalConfig, *, use_dask: bool) -> Self:
+            np.roll(np.arange((new_front := n_center + n_front)), n_front),
+            np.arange(new_front, new_front + n_back),  # rest of vars
+        )
+    )
```
Not sure I'm reading this correctly, but is this equivalent to `order = [all_vars.index(v) for v in [vars_with_dim + vars_without_dim]]`? I'm a bit confused why we don't end up using the values of front/back right now (just their lengths).

alxmrs replied:
No, it's not equivalent to that. For one, `all_vars` doesn't include the (likely) generated combinations of new vars produced in a rearrage operation (e.g. `time (variable lev) lat lon`). For example, when I step in my IDE, `all_vars`, `vars_with_dim`, and `vars_without_dim` only ever include values `['uo', 'vo', 'thetao', 'so', 'zos']`; however, `data_with_dim` will have these variables cross the number of levels. 

This is confusing! So, to help, I'm adding an example in comments.

jder commented on src/ocean_emulators/utils/data.py:
```diff
@@ -393,7 +559,11 @@ def _static_data_checks(data):
-    def _rename(data, means, stds):
+    if src.is_compact:
+        src_ = src.map_data(with_lat_lon_coords)
+        return compute_anomalies(src_, anomalies_vars)
```
:sheep: Might be slightly nicer to compute anomalies once at the end of this function and put the validation in an else rather than an early return?

jder commented on tests/conftest.py:
```diff
@@ -269,6 +270,32 @@ def decode(cls, da: xr.DataArray) -> tuple[Self, NDArray[np.uint]]:
+        data = data.drop_vars(vars_)
+
+    return data
+
```
(This is a good example of why I don't like a separate scripts directory; this is verbatim copied from there, I think?)

alxmrs replied:
Yes, that's right -- and, a fair point.

jder commented on tests/test_utils_data.py:
```diff
@@ -242,7 +242,7 @@ def data_init(hist: int):
         data_std = data.std() * 0.0 + 1.0
         src = DataSource("test", data, data_mean, data_std)
-        val = validate_data(src)
+        val = validate_data(src, [])
```
I think `tensor_map.boundary_var_names` might be less fragile than hard-coding []?

alxmrs replied:
I see what you mean now! I didn't realize that we already init tensor map in the test.

jder commented on tests/test_stepper.py:
```diff
@@ -64,7 +64,7 @@ def inf_data_init(hist: int):
         data_std: xr.Dataset = data.std() * 0.0 + 1.0
         test_data = DataSource("test-data", data, data_mean, data_std)
-        val = validate_data(test_data)
+        val = validate_data(test_data, [])
```
(same comment here about using tensor_map for boundary var names)

alxmrs replied:
I'm passing in the argument in order to reduce dependency on tensor_map (as a gradual refactor).

jder replied:
I like the fact that you're passing in the argument, I'm just suggesting that you pass in `tensor_map.boundary_var_names` (tensor_map is defined earlier on in this function) rather than `[]`.

jder commented on tests/test_datasets.py:
```diff
@@ -393,6 +401,59 @@ def test_new_loaders__are_equal_to_v1_data_loader(train_config, loader_version):
+
+            assert not np.any(x_not_equal), (
+                f"{len(x_not_equal_index)} values differ: "
+                f"{x_orig[x_not_equal_index[0]]} != {x_new[x_not_equal_index[0]]}."
```
So I think this is exactly identical to the code above except you've got a `[0]` here. Is this a meaningful difference or should we make a `compare_loaders` function which deduplicates this code?

jder commented on tests/test_datasets.py:
```diff
@@ -393,6 +401,59 @@ def test_new_loaders__are_equal_to_v1_data_loader(train_config, loader_version):
+
+    with (
+        make_loader(flat_config, version=LoaderVersion.OM4_TORCH) as flat_loader,
+        make_loader(compact_config, version=LoaderVersion.OM4_TORCH) as compact_loader,
```
Reading this again, I think this is broken -- we can't have both multiton scopes active at once. We need to use `with make_loader(flat…): original_samples = …` and then separately `with make_loader(compact…): new_samples = …`

alxmrs replied:
I'd like to understand this better -- mind if we jump on a call? Right now, the only multiton we create is the TensorMap; the rest of the relevant values are injected in based on the config/params:

https://github.com/suryadheeshjith/Ocean_Emulator/blob/ff81c07df95dda17ba1dfb9971412c08c781c751/tests/test_datasets.py#L65

jder reviewed (commented):
Thanks! Just a few questions/comments from me. I'm ready to approve once you've got some benchmarks we can look at. I do think this is adding a lot of duplication, but I hope we can delete that (and conditional_rearrange and so on) after we remove the non-compact form of data.

alxmrs committed:
Commenting to make logic clearer.

alxmrs committed:
Favoring if block instead of returning early in order to deduplicate last step.

alxmrs committed:
Function is not copied, but imported into script. Now, `clone_data.py` depends on `ocean_emulators`.

alxmrs committed:
Extracted common assert function.

alxmrs committed:
Better to use tensor map args instead of an empty list.

alxmrs committed:
Fix potential cause of subtle bug with context managers in tests.

alxmrs committed:
Added comments about implementation details.

jder commented on scripts/clone_data.py:
```diff
@@ -11,20 +11,25 @@
 from dask.distributed import LocalCluster
 from tenacity import retry
 
+from ocean_emulators.utils.data import compact_dataset
```
<3

jder commented on tests/test_datasets.py:
```diff
@@ -361,36 +364,74 @@ def test_inference__data_is_not_zero(inference_loader_pair):
+    with make_loader(compact_config, version=LoaderVersion.OM4_TORCH) as compact_loader:
+        new_samples = [extract_sample_arrays(sample) for sample in compact_loader]
+
+    assert_equal_samples(original_samples, new_samples)
```
Beautiful 🤩

jder reviewed (commented):
Continues to LGTM! Thanks for the fixes, happy to merge when we've got the benchmarks sorted out!

jder commented:
@alxmrs did anything change here since my last review? Maybe we can chat live when you're around?

jder reviewed (approved):
Discussed live that we should merge this to avoid having it be on a branch forever and we'll time box follow ups on identifying the performance wins/losses here.
